twilight-http-ratelimiting/src/in_memory/global_bucket.rs

```diff
@@ -0,0 +1,121 @@
+//! Bucket implementation for a global ratelimit.
+
+use super::bucket::BucketQueue;
+use crate::RatelimitHeaders;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::Instant;
+
+/// seconds per period
+const PERIOD: u64 = 1;
+/// requests per period
+const REQUESTS: u32 = 50;
+
+/// Global bucket. Keeps track of the global rate limit.
+#[derive(Debug, Clone)]
+pub struct GlobalBucket(Arc<InnerGlobalBucket>);
+
+impl GlobalBucket {
+    /// Creates a new global bucket using custom ratelimit values.
+    ///
+    /// `period` is given in seconds.
+    ///
+    /// `requests` indicates the amount of requests per period.
+    #[must_use]
+    pub fn with_ratelimit(period: u64, requests: u32) -> Self {
+        Self(InnerGlobalBucket::new(period, requests))
+    }
+
+    /// Queue of global ratelimit requests.
+    pub fn queue(&self) -> &BucketQueue {
+        &self.0.queue
+    }
+
+    /// Whether the global ratelimit is exhausted.
+    pub fn is_locked(&self) -> bool {
+        self.0.is_locked.load(Ordering::Relaxed)
+    }
+}
+
+impl Default for GlobalBucket {
+    fn default() -> Self {
+        Self(InnerGlobalBucket::new(PERIOD, REQUESTS))
+    }
+}
+
+/// Inner struct to allow [`GlobalBucket`] to return an [`Arc`].
+#[derive(Debug)]
+struct InnerGlobalBucket {
+    /// Queue to receive rate limit requests.
+    pub queue: BucketQueue,
+    /// currently waiting for capacity.
+    is_locked: AtomicBool,
+}
+
+impl InnerGlobalBucket {
+    /// Creates a new bucket and starts a task processing incoming requests.
+    fn new(period: u64, requests: u32) -> Arc<Self> {
+        let this = Self {
+            queue: BucketQueue::default(),
+            is_locked: AtomicBool::default(),
+        };
+        let this = Arc::new(this);
+
+        tokio::spawn(run_global_queue_task(this.clone(), period, requests));
+
+        this
+    }
+}
+
+#[tracing::instrument(name = "background global queue task", skip_all)]
+async fn run_global_queue_task(bucket: Arc<InnerGlobalBucket>, period: u64, requests: u32) {
+    let mut time = Instant::now();
+
+    while let Some(queue_tx) = bucket.queue.pop().await {
+        wait_if_needed(bucket.as_ref(), &mut time, period, requests).await;
+
+        let ticket_headers = if let Some(ticket_headers) = queue_tx.available() {
+            ticket_headers
+        } else {
+            continue;
+        };
+
+        if let Ok(Some(RatelimitHeaders::Global(headers))) = ticket_headers.await {
+            tracing::debug!(seconds = headers.retry_after(), "globally ratelimited");
+
+            bucket.is_locked.store(true, Ordering::Release);
+            tokio::time::sleep(Duration::from_secs(headers.retry_after())).await;
+            bucket.is_locked.store(false, Ordering::Release);
+        }
+    }
+}
```

Comment:
Let's take a look at the flow of the logic here:

1. global queue task waits for an item in the queue, which happens on every ratelimiter request
2. the global ratelimiter waits if needed
3. the global queue notifies the other half they can send the request
  a. if the other half is still around, the global queue receives a receiving half to receive headers from
  b. otherwise, the global queue can know the other half doesn't exist and loops over to the next request
4. the global queue waits for the headers from the other half; if the headers are of a global ratelimit then it sleeps, causing the entire system to stall, as intended
5. it loops over

In terms of preventing more than a configured number of requests in a given period, this system of preemptive sleeping works great. The problem here is that this is synchronous: only one request will be processed at a time. Practically speaking, this means the global bucket won't limit to, say, 50 requests/1 second, but rather to a much smaller number of requests because the global bucket is a bottleneck between each request. Requests to ping Discord take an average latency of 102ms right now [according to the status page](https://discordstatus.com/), so functionally the global bucket will bottleneck users to < 10 requests/second, if not much fewer depending on the routes in use.

We need to change the implementation here to solve this. First instinct is that hitting a global ratelimit header with this system is unlikely with the preemptive ratelimiter in place, so what we could do instead is spawn a task for each set of headers while waiting. With a mutex in place instead of the `is_locked` boolean, we could lock up the ratelimiter for the duration. That works.

But this has its own flaw: if a user makes, say, 100 requests to the ratelimiter with a global bucket of 50 requests/1 second, then across just over 1 second they'll receive all 100 request allotments. If requests don't all take the same time (very likely), or they get stalled, or something else, then users could fire off more than 50 requests in a given second and still hit the global ratelimit. It's also impossible to determine when Discord will count against a request against a given second. We can see how this could fail under this circumstance:

1. User requests 100 ratelimiter tickets for requests, current configuration is 50 requests/second
2. 50 tickets are granted in the first second
3. 40 are processed by Discord in that second
4. 50 tickets are granted in the second second
5. those 50, plus the 10 from the previous second, are processed by Discord
6. that was 60 in a second -- oops, global ratelimit hit

The solution I see is to simultaneously limit the number of requests allotted in a given second *and* limit the number of in-flight requests to the configured number of requests. We can see why this prevents the previous scenario:


1. User requests 100 ratelimiter tickets for requests, current configuration is 50 requests/second
2. First second
  1. 50 tickets are granted in the first second and their requests are sent to Discord
  2. 40 are processed by Discord in that second
  3. 30 requests get responses and are processed by the global queue
3. Second.. second
  1. Because 20 requests are still in flight and their global ratelimit status is indeterminate, 30 tickets are granted
  2. the responses of the 10 requests that were processed by Discord but didn't receive responses are received
  3. the 10 requests that weren't processed by Discord last second are processed and return responses
  4. 20 of the requests from this second are received
  5. 10 are not, so there are 10 requests still in flight
4. Third second
  1. because 10 requests are in flight, there are 40 tickets allotted for the second
  2. the cycle continues
5. it's mathematically and logically impossible for a global ratelimit

To accomplish this, my suggestion is we use a combination of tracking the time (like now), spawning tasks to process responses, *and* a Tokio semaphore with the configured number of requests as allotments. When one second passes over into another, the first processed request of that second can add allotments for the number of responses that were received last second. If you want tips on how to implement this let me know and I can provide them, it's a bit complicated.

Let me know if I'm thinking about this wrong, but this implementation of using a dual timer and semaphore seems like the *very correct* way of doing this.

Reply:
The current time tracker is inefficient, `wait_if_needed` will ~always sleep and yield one request/(period/requests) and therefore evenly space them out. We could either use a leaky bucket algorithm (we can take a dependency upon a crate implementing it) or mirror the implementation of `twilight_gateway::CommandRatelmitier`. The ratelimiter in the gateway was quite tricky, so I'd recommend adding lots of tests for this module.

Reply:
Are you sure that Discord resets every second opposed to evenly allowing more? In that case this way would allow requests to complete earlier.

Reply:
> Are you sure that Discord resets every second opposed to evenly allowing more? In that case this way would allow requests to complete earlier.

I misspoke, thanks for the catch. I don't believe the docs explicitly state this, but it is a GCRA implementation, so it's actually a leaky bucket that will replenish over time.